<a href="https://colab.research.google.com/github/rahiakela/grokking-deep-learning/blob/3-forward-propagation/forward_propagation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# A Simple Neural Network Making a Prediction

## Step 1: Predict
You learned about the paradigm predict, compare, learn. In this guide, we’ll dive deep into the first step: **predict**.

you’ll learn more about what these three different parts of a neural network
prediction look like under the hood. Let’s start with the first one: the **data**.

Later, you’ll find that the number of datapoints you process at a time has a significant
impact on what a network looks like. You might be wondering, “How do I choose how
many datapoints to propagate at a time?” The answer is based on whether you think the
neural network can be accurate with the data you give it.

For example, if I’m trying to predict whether there’s a cat in a photo, I definitely need to
show my network all the pixels of an image at once. Why? Well, if I sent you only one
pixel of an image, could you classify whether the image contained a cat? Me neither!

That’s a general rule of thumb, by the way: always present enough information to the
network, where “enough information” is defined loosely as how much a human might
need to make the same prediction.

As it turns out, you can create a network only after
you understand the shape of the input and output datasets (for now, shape means “number
of columns” or “number of datapoints you’re processing at once”).

Now that you know you want to take one input datapoint and output one prediction, you
can create a neural network. Because you have only one input datapoint and one output
datapoint, you’re going to build a network with a single knob mapping from the input point
to the output. (Abstractly, these “knobs” are actually called weights) 

So, without further ado, here’s your first neural network, with a
single weight mapping from the input “# toes” to the output “win?”:

### An empty network

In [0]:
weight = 0.1

def neural_network(input, weight):
  # Multiplying input by weight
  prediction = input * weight
  return prediction

### Inserting one input datapoint and Depositing the prediction

In [8]:
number_of_toes = [8.5, 9.5, 10, 9]

input = number_of_toes[0]
pred = neural_network(input, weight)
print(pred)

0.8500000000000001


You just made your first neural network and used it to predict! Congratulations! It should be 0.85. So what is a neural network? For now, it’s one or
more weights that you can multiply by the input data to make a prediction.

### How does the network learn?

Trial and error! First, it tries to make a prediction. Then, it sees whether the prediction was too
high or too low. 

Finally, it changes the weight (up or down) to predict more accurately the
next time it sees the same input.

### What does this neural network do?

#### It multiplies the input by a weight. It “scales” the input by a certain amount.

In the previous section, you made your first prediction with a neural network. A neural network,
in its simplest form, uses the power of multiplication. It takes an input datapoint (in this case,
8.5) and multiplies it by the weight. If the weight is 2, then the neural network will double the
input. If the weight is 0.01, then the network will divide the input by 100. As you can see, some
weight values make the input bigger, and other values make it smaller.


#### Conclusion

The interface for a neural network is simple. It accepts an input variable as information and a
weight variable as knowledge and outputs a prediction. 

Every neural network you’ll ever see
works this way. It uses the knowledge in the weights to interpret the information in the input
data. 

Later neural networks will accept larger, more complicated input and weight values, but
this same underlying premise will always ring true.

Notice several things. 

* First, the neural network does not have access to any information
except one instance. 
* If, after this prediction, you were to feed in number_of_toes[1], the
network wouldn’t remember the prediction it made in the last timestep. 
* A neural network
knows only what you feed it as input. It forgets everything else. 

Later, you’ll learn how to
give a neural network a “short-term memory” by feeding in multiple inputs at once.

Another way to think about a neural network’s weight value is as a measure of sensitivity
between the input of the network and its prediction. 
* If the weight is very high, then even the
tiniest input can create a really large prediction! 
* If the weight is very small, then even large
inputs will make small predictions. 

This sensitivity is akin to volume. “Turning up the weight”
amplifies the prediction relative to the input: weight is a volume knob!

Note that neural networks don’t predict just positive numbers—they can also predict negative
numbers and even take negative numbers as input. Perhaps you want to predict the probability
that people will wear coats today. If the temperature is –10 degrees Celsius, then a negative
weight will predict a high probability that people will wear their coats.

## Making a prediction with multiple inputs